docs/rfc/rfc-017-abci++-vote-extension-propag.md

```diff
@@ -0,0 +1,438 @@
+
+The blocksync reactor featuring light client verification is being actively worked on (tentatively
+for v0.37). So it is best if this RFC does not try to solve that problem, but just makes sure
+the outcomes are compatible with that effort (TODO: check this paragraph with Callum/Jasmina).
+
+In subsection [Cases to Address](#cases-to-address), we concluded that we can focus on
+solving case (g) in theortical terms.
+However, as the current Tendermint version does not yet support switching back to blocksync once a
+node has switched to consensus, we need to split case (g) into two cases. When a full node needs to
+catch up...
+
+- **(g.1)** ... it has not switched yet from the blocksync reactor to the consensus reactor, or
+
+- **(g.2)** ... it has already switched to the consensus reactor.
+
+This is important in order to discuss the different possible implementations.
+
+#### Base Implementation: Persist and Propagate Extended Commit History
+
+In order to circumvent the fact that we cannot switch from the consensus reactor back to blocksync,
+rather than just keeping the few (TODO: 1?, 2?) most recent extended commits, nodes will need to keep
+and gossip a backlog of extended commits so that the consensus reactor can still propose and decide
+in out-of-date heights.
+
+The base implementation (for which an experimental patch exists) consists in the conservative
+approach of persisting in the block store *all* extended commits for which we have also stored
+the full block.
+Note that this does not apply to light blocks that might have been saved in the block store as a
+result of running statesync at startup.
+
+Then, we modify the blocksync reactor so that peers *always* send requested full blocks together
+with the corresponding extended commit in the `BlockResponse` messages. This guarantees that the
+block store being reconstructed by blocksync has the same information as that of peers that are
+up to date.
+Thus, blocksync has all the data it requires to switch to the consensus reactor, as long as:
+
+- The node is still at height 0 (where no commit or extended commit is needed)
+- The node has processed at least 1 block in blocksync
+
+Additionally, when a validator falls behind while having already switched to the consensus reactor,
+peer nodes have enough information in their block store to reconstruct the extended votes and send
+them to the validator falling behind.
+
+This solution requires a few changes to the consensus reactor:
+
+- upon saving block for a given height in the block store at decision time, save the extended commit
+  as well
+- in the catch-up mechanism, when a node realizes that another peer is more than 2 heights
+  behind, it uses the extended commit (rather than the canoncial commit as done previously) to
+  reconstruct the votes with their corresponding extensions
+
+The changes to the blocksync reactor are more substantial:
+
+- the `BlockResponse` message is extended to include the extended commit
```

Comment:
Can you clarify which extended commit you mean here, from height `h-1`, `h`, or `h-1`?

Reply:
Good point